hack/verify-go-versions.sh

```diff
@@ -17,14 +17,19 @@
 set -e
 set -o pipefail
 
-VERSION=$(grep "go 1." go.mod | sed 's/go //')
+MINIMAL_VERSION=$(grep "go 1." go.mod | sed 's/go //')
 
-# grep "FROM golang:" Dockerfile | { ! grep -v "${VERSION}"; } || { echo "Wrong go version in Dockerfile, expected ${VERSION}"; exit 1; }
-# grep -w "go-version:" .github/workflows/*.yaml | { ! grep -v "go-version: v${VERSION}"; } || { echo "Wrong go version in .github/workflows/*.yaml, expected ${VERSION}"; exit 1; }
+# grep "FROM golang:" Dockerfile | { ! grep -v "${MINIMAL_VERSION}"; } || { echo "Wrong go version in Dockerfile, expected ${MINIMAL_VERSION}"; exit 1; }
+# grep -w "go-version:" .github/workflows/*.yaml | { ! grep -v "go-version: v${MINIMAL_VERSION}"; } || { echo "Wrong go version in .github/workflows/*.yaml, expected ${MINIMAL_VERSION}"; exit 1; }
 # Note CONTRIBUTING.md isn't copied in the Dockerfile
 # if [ -e CONTRIBUTING.md ]; then
-#   grep "golang.org/doc/install" CONTRIBUTING.md | { ! grep -v "${VERSION}"; } || { echo "Wrong go version in CONTRIBUTING.md expected ${VERSION}"; exit 1; }
+#   grep "golang.org/doc/install" CONTRIBUTING.md | { ! grep -v "${MINIMAL_VERSION}"; } || { echo "Wrong go version in CONTRIBUTING.md expected ${MINIMAL_VERSION}"; exit 1; }
 # fi
+
 if [ -z "${IGNORE_GO_VERSION}" ]; then
-  go version | { ! grep -v go${VERSION}; } || { echo "Unexpected go version installed, expected ${VERSION}. Use IGNORE_GO_VERSION=1 to skip this check."; exit 1; }
+  ENV_VERSION=$(go version | awk '{print $3}' | sed 's/go//')
+  if [[ "$ENV_VERSION" < "$MINIMAL_VERSION" ]]; then
```

Comment:
I think it is safer to insist on a positive outcome from test for an acceptable relationship than accept a negative outcome from a test for an unacceptable relationship --- the latter can accept unexpected failures.

Reply:
I tend to disagree with this comment.

if you had to write a function with your custom logic and had to verify a condition before starting, how would you write your function?

![image](https://github.com/kubestellar/kubestellar/assets/19717747/c14f9f10-b697-417a-b3ed-feb4d6fa7bfc)

Or:
![image](https://github.com/kubestellar/kubestellar/assets/19717747/ce37c6f3-ceb4-4367-b4c8-b277b13e77b1)

of course it is a matter of code styling and both options will work, but best practices definitely mention the first option as the preferred approach. it reduces code indentation which results in a more readable and maintainable code and it is easier for understanding.

another point - the original code of this script was checking negative test (pay attention for the NOT) so I'm not sure what exactly has changed. 
also adding here another example from a well known CNCF incubation project, KubeEdge, which also checks go version and does it using negative test. see example here:
https://github.com/kubeedge/kubeedge/blob/master/hack/lib/golang.sh#L35-#L43.

I can easily add a check that verifies go is installed before the versions comparison if you think it's safer.
other than that, I don't really see which unexpected failures may happen and I'm happy to hear if you have any example.

Reply:
My point is not about indentation (you can get the indentation you like with either sort of test by appropriate engineering on the script) but about the ways that a command in `bash` can have a non-0 exit code. Bash has ways to fail that you typically do not see in programming languages like Go. In the case at hand, what if `$MINIMAL_VERSION` did not get set to the expected string but rather the empty string?

In the older script, the test `grep -v go${VERSION}` is testing for the acceptable condition; any failure to demonstrate the acceptable condition is treated as a problem.

Reply:
That is incorrect.
In older script, if `{VERSION}` is empty, the script will return successfully which is a problem (is it?).
I think the main question here is - **what is this script trying to verify**??

IMO we should make a clear distinction between whatever is part of KS code, and the environment of the machine which tries to run KS locally. 
**This script is NOT checking KS code, it's checking the machine environment.** 
this was the situation in older script and in this new script.

go.mod and this verify script are both part of KS code. 
if a user clones the repo and starts playing with `go.mod` file, he shouldn't be surprised if things breaks.
the assumption should be that in the older script the var ${VERSION} contains the correct version as written in `go.mod` (in newer script nothing has changed, var just renamed). 

if one of KS team members will change `go.mod` in a way that break things, the PR will fail and we will catch it.

this script only checks the machine env and reports if the env doesn't match the minimal requirements of KS.

I've added another check to verify that go is installed on the machine to verify we don't get garbage.

Reply:
> if {VERSION} is empty, the script will return successfully

Oh, good point. That is indeed a bug, unless something else is failing when go is not installed (which I think was already true, but no harm in adding an explicit check).